sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ValuePostprocessor.scala

```diff
@@ -218,4 +220,68 @@ final class ValueEnricher(
     enrichTransaction(incompleteTx.transaction).map(transaction =>
       incompleteTx.copy(transaction = transaction)
     )
+
+  /*
+    This will renormalize a contract w.r.t a (up/down)grade template.
+    In practice this potentially
+     - adds/removes none fields
+     - change the transaction version.
+   */
+  private[this] def reversionValue(
+      typ: Ast.Type,
+      value: Value,
+      dstVer: TransactionVersion,
+  ): Result[Value] =
+    translateValue(typ, value).map(_.toNormalizedValue(dstVer))
+
+  // This
+  def reversion(
+      contract: FatContractInstance,
+      dstTmplId: Ref.TypeConName,
+  ): Result[FatContractInstance] = {
+    if (contract.templateId == dstTmplId)
+      ResultDone(contract)
+    else
+      for {
+        dstPkg <- handleLookup(compiledPackages.pkgInterface.lookupPackage(dstTmplId.packageId))
+        _ <-
+          if (
+            dstPkg.pkgName == contract.packageName && dstTmplId.qualifiedName == contract.templateId.qualifiedName
+          )
+            Result.unit
+          else
+            ResultError(
+              Error.Postprocessing.ReversioningMismatch(
+                contract.contractId,
+                dstPkg.pkgName,
+                dstTmplId.qualifiedName,
+                contract.packageName,
+                contract.templateId.qualifiedName,
+              )
+            )
+        dstTxVersion = TransactionVersion.assignNodeVersion(dstPkg.languageVersion)
+        dstArgType = Ast.TTyCon(dstTmplId)
+        createArg <- reversionValue(dstArgType, contract.createArg, dstTxVersion)
+        keyOpt <- contract.contractKeyWithMaintainers match {
+          case Some(GlobalKeyWithMaintainers(globalKey, maintainers)) =>
+            for {
+              keyDef <- handleLookup(compiledPackages.pkgInterface.lookupContractKey(dstTmplId))
+              keyVal <- reversionValue(keyDef.typ, globalKey.key, dstTxVersion)
+            } yield Some(
+              GlobalKeyWithMaintainers.assertBuild(
+                dstTmplId,
+                keyVal,
+                maintainers,
+                dstPkg.pkgName,
+              )
+            )
+          case None => ResultNone
+        }
+      } yield contract.toImplementation.copy(
```

Comment:
Previously that argument was consistent with the templateId so it is not nice this is no longer the case. It also means that `(contract.templateId == dstTmplId) => ResultDone(contract)` will not work for pre-converted contracts. An alternative could be to also template the `templateId` with the converted templateId. If this is to be considered `Hash.assertHashContractInstance` would need to be modified not to include the `packageId`.

Reply:
@remyhaemmerle-da Please read @simonmaxen-da 's first comment and double-check there is no fundamental problem with the engine. That said, it is totally up to you how you handle things inside of the engine.

> Maybe add an optional reversionedPackageId to FatContractInstance ?

Please keep in mind that such a change may be challenging to consume in the Canton repo. If you add the field, please take care of the fall-out, too.